### ISSUE TITLE
Problem iterating through container/List

### ISSUE THREAD

**Issue by jpaulm** [jpaulm/gofbp#11]
I was trying to maintain a list of `*Process` pointers, which I would later walk through to start the processes.  As far as I can see, when I iterate through the list (using Next, etc.), it retrieves each entry as an `interface{}`, which I should be able to turn back into a  `*Process`, but it is returned as a `Process`...   which I couldn't turn back into a pointer.

It may be my bug, but after spending several hours on it, I followed someone's advice and used a slice instead... and this worked fine!  Except... that you have to choose a capacity, and if I pick a big one, I believe this takes up space.  List would be more elegant!

**Comment by egonelbre** [comment]
The reason the `list.List` idea didn't work, was because `PushFront` was using a dereference.

```
proc := &Process{}
var procs list.List
procs.PushFront(*proc) // should be `procs.PushFront(proc)`
```

However, `list.List` usage is generally discouraged, because it's often slower and easier to get slightly wrong. So, slices is the preferred usage.

With regards to slices - they do not require specifying the length, it's an optional feature and used only for performance reasons and when you know upfront the exact length. You can add items to a slice with append:

```
var procs []*Process
procs = append(procs, proc)

for _, proc := range procs {
    fmt.Println(proc)
}
```

With `append` the runtime adds item to the slice when there is capacity, or reallocates a new underlying buffer when necessary. Blog post https://go.dev/blog/slices describes them in depth, if you are interested.

**Comment by jpaulm** [comment]
Odd!  I don't remember closing this!  Plus, my list was not a list of procs - it was supposed to be a list of *procs...!  The problem arose when I was iterating through it using Next - Next returns an interface{} which insisted on dereferencing the *proc , turning it into a proc!  Too much trouble to dig into some more!

**Comment by egonelbre** [comment]
> Odd! I don't remember closing this!

Ah, that came from my commit message. If it contains `Fixes #123` it ends up automatically closing the corresponding issue 123.

In general `Next` returns the same item with the same type as argument to `PushFront` or `PushBack`.

**Comment by jpaulm** [comment]
Going to close this - too much work to create a test case, etc. :-)
